apollo-router-core/src/services/http_compat.rs

```diff
@@ -2,12 +2,76 @@
 //!
 //! To improve their usability.
 
+use derive_more::From;
+use multimap::MultiMap;
 use std::{
     cmp::PartialEq,
     hash::Hash,
     ops::{Deref, DerefMut},
 };
 
+/// Temporary holder of header name while for use while building requests and responses. Required
+/// because header name creation is faillable.
+#[derive(From, Eq, Hash, PartialEq)]
+pub enum IntoHeaderName {
+    String(String),
+    HeaderName(HeaderName),
+}
+
+/// Temporary holder of header value while for use while building requests and responses. Required
+/// because header value creation is faillable.
+#[derive(From, Eq, Hash, PartialEq)]
+pub enum IntoHeaderValue {
+    String(String),
+    HeaderValue(HeaderValue),
+}
+
+impl From<&str> for IntoHeaderName {
+    fn from(name: &str) -> Self {
+        IntoHeaderName::String(name.to_string())
+    }
+}
+
+impl From<&str> for IntoHeaderValue {
+    fn from(value: &str) -> Self {
+        IntoHeaderValue::String(value.to_string())
+    }
+}
+
+impl From<&String> for IntoHeaderName {
+    fn from(name: &String) -> Self {
+        IntoHeaderName::String(name.to_owned())
+    }
+}
+
+impl From<&String> for IntoHeaderValue {
+    fn from(value: &String) -> Self {
+        IntoHeaderValue::String(value.to_owned())
+    }
+}
+
+impl TryFrom<IntoHeaderName> for HeaderName {
+    type Error = http::Error;
+
+    fn try_from(value: IntoHeaderName) -> Result<Self, Self::Error> {
+        Ok(match value {
+            IntoHeaderName::String(name) => HeaderName::try_from(name)?,
+            IntoHeaderName::HeaderName(name) => name,
+        })
+    }
+}
+
+impl TryFrom<IntoHeaderValue> for HeaderValue {
+    type Error = http::Error;
+
+    fn try_from(value: IntoHeaderValue) -> Result<Self, Self::Error> {
+        Ok(match value {
+            IntoHeaderValue::String(value) => HeaderValue::try_from(value)?,
+            IntoHeaderValue::HeaderValue(value) => value,
+        })
+    }
+}
```

Comment:
This is very much a "style/taste" kind of thing, so I'd be interested to read what other people thought about this and I don't think that there is a single answer. So: more feedback required, but my feedback is:

Stylistically, this seems unusual to me. Don't HeaderValue and HeaderName already implement TryFrom for AsRef<str>? Also, if you were doing this, I'd look at using AsRef<str> rather than &String, etc...

Reply:
I hadn't realized until recently, but the native `http::Request`/`http::Response` builder does something similar:
https://github.com/hyperium/http/blob/master/src/request.rs#L925

`TryInto` is not called before the call to `build()`. That way the user can pass in strings as headers, and the call to build will cause the fallible conversions to happen all at once.